File: 102. Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
//     private List<List<Integer>> res;
    
//     public List<List<Integer>> levelOrder(TreeNode root) {
//         List<List<Integer>> res = new ArrayList<>();
//         int ht = height(root);
//         for (int i = 0; i < ht; i++) {
//             List<Integer> temp = rec(root,i, new ArrayList<Integer>());
//             res.add(new ArrayList<>(temp));
//         }
//         return res;
//     }
    
//     public List<Integer> rec(TreeNode root, int level, List<Integer> temp) {
//         //List<Integer> temp = new ArrayList<>();
//         if (root == null) return temp;
//         if (level == 0) {
//             temp.add(root.val);
//             return temp;
//         };
//         rec(root.left, level - 1, temp);
//         rec(root.right, level - 1, temp);
//         return temp;
//     }
    
//     public int height(TreeNode root) {
//         if (root == null) return 0;
//         return Math.max(height(root.left),height(root.right)) + 1;
//     }
    
    // Use queue
    // public List<List<Integer>> levelOrder(TreeNode root){
    //     if (root == null) return new ArrayList<>();
    //     List<List<Integer>> res = new ArrayList<>();
    //     Queue<TreeNode> q = new LinkedList<>();
    //     q.add(root);
    //     while (!q.isEmpty()) {
    //         List<Integer> oneLevel = new ArrayList<>();
    //         int size = q.size();
    //         for (int i = size; i > 0; i--) {
    //             TreeNode t = q.peek();
    //             oneLevel.add(t.val);
    //             q.poll();
    //             if (t.left != null) {
    //                 q.add(t.left);
    //             }
    //             if (t.right != null) {
    //                 q.add(t.right);
    //             }
    //         }
    //         res.add(new ArrayList<>(oneLevel));
    //     }
    //     return res;
    // }
    
    private List<List<Integer>> res = new ArrayList<>();
    // Use recursion 
    public List<List<Integer>> levelOrder(TreeNode root){
        if (root == null)
            return new ArrayList<>();
        oneLevel(root, 0, res);
        return res;
    }
    
    public void oneLevel(TreeNode root, int level, List<List<Integer>> res) {
        if (res.size() == level)
            res.add(new ArrayList<>());
        res.get(level).add(root.val);
        if (root.left != null)
            oneLevel(root.left,level + 1, res);
        if (root.right != null)
            oneLevel(root.right, level + 1, res);
    }
}

File: amazon/python/102. Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrder(self, root):
        levels = []
        if not root:
            return levels
        
        def helper(node, level):
            if len(levels) == level:
                levels.append([])
            
            levels[level].append(node.val)
            
            if node.left:
                helper(node.left, level + 1)
            if node.right:
                helper(node.right, level + 1)
        
        helper(root,0)
        return levels 
    
    ## Time:
    ## Space: 
    
    
#     def levelOrder(self, root):
#         """
#         :type root: TreeNode
#         :rtype: List[List[int]]
#         """
#         res = []
#         queue = []
#         temp_queue = []
        
#         if not root:
#             return res 
#         queue.append(root)
        
#         while queue != []:
#             size = len(queue)
#             for i in range(size):
#                 cur = queue.pop(0)
#                 temp_queue.append(cur.val)
#                 if cur.left:
#                     queue.append(cur.left)
#                 if cur.right:
#                     queue.append(cur.right)
#             res.append(temp_queue)
#             temp_queue = []
        
#         return res 
    # Time: O(n)
    # Space: O(n)
                
            
